### DISCUSSION TITLE
Added images should have meaningful or practical filenames

### DISCUSSION THREAD

**Discussion by djalilhebal** [GitJournal/GitJournal#693] [Feature Requests]
## Environment

**App version: GitJournal 1.75.1+2324**
**Device information:**
 - OS name and version: Android 8.1 (Go edition)
 - Manufacturer: Samsung
 - Model: SM-J410F

## Description

GitJournal adds images with random filenames (`image_picker{random number}.{extension}`).

**What you'd like to happen:**
Added images should retain their original filenames.

**Alternatives you've considered:**
Files should be prefixed with the note's ID/name.  
For example, if I am editing `2021...42.md` and add an image, its name should be `2021...42-image-1.png`. The next one should be named `2021...42-image-2.png`, and so on.

**Images:**
![VS Code screenshot](https://user-images.githubusercontent.com/32184973/112013815-a222a980-8b2a-11eb-9482-bf89e469917c.png)

**Comment by ignon** [comment]
Would love to see this implemented. An ideal solution would probably be a popup text input with the original filename as a default value (or prefixed if set in options)

**Comment by majerich** [comment]
Can this be implemented as a configuration setting? I actually prefer having plain hex file names when they're attached to notes. Using a plug-in in Obsidian to provide just this ability.

**Comment by vHanda** [comment]
> Can this be implemented as a configuration setting? I actually prefer having plain hex file names when they're attached to notes. Using a plug-in in Obsidian to provide just this ability.

I was planning on making it configurable. Here are the current options in my mind -

* Follow the file name of the note. Add _1, _2, etc in the case of a conflict.
* Retain the original filenames, again _1, _2 in the case of a conflict.
* A random name. @majerich : Could you please give me the exact format + length used by the Obsidian plugin. I may as well follow it.
* Show a dialog each time.

Could you please let me know if there is any other option you would like?

Note to self: We don't have an original name for new photos. Maybe have a different setting for new images?

**Comment by vHanda** [comment]
Sigh, it seems that getting the actual image's filename isn't yet implemented - https://github.com/flutter/flutter/issues/64685 - pfff, I'll have to first implement that in both Android and iOS. This has now become a much larger task.

**Comment by majerich** [comment]
> > Can this be implemented as a configuration setting? I actually prefer having plain hex file names when they're attached to notes. Using a plug-in in Obsidian to provide just this ability.
> 
> I was planning on making it configurable. Here are the current options in my mind -
> 
>     * Follow the file name of the note. Add _1, _2, etc in the case of a conflict.
> 
>     * Retain the original filenames, again _1, _2 in the case of a conflict.
> 
>     * A random name. @majerich : Could you please give me the exact format + length used by the Obsidian plugin. I may as well follow it.
> 
>     * Show a dialog each time.
> 
> 
> Could you please let me know if there is any other option 

So the Obsidian community plug-in is named "Unique attachments". Current version is 1.0.3  when it uniquely renames a file, it retains the file's original extention but replaces the name with a random, 32 character hexadecimal value. I'm attaching a screen shot example.

![20210916_100752](https://user-images.githubusercontent.com/89341463/133638578-dce371df-e8f3-4e98-a600-a27218bdb725.jpg)

Supports 5 configurable options:
1. File types to rename (comma separated list of extensions you want it to handle. png,jpg,gif are the defaults)
2. Ignore folders (a new line separated list of folder names with relative paths to ignore even if they contain pattern matched extentions)
3. Rename only linked attachments Boolean (if an image is not linked in the current note, it will not be renamed)
4. Save a previous name Boolean (when it changes the file's name, it uses the original name as the link's description
5. Delete duplicates Boolean (if there are duplicate files in the same folder but with different names, it will delete the duplicates. If turned off, no file's will be renamed at all on this condition)

Hope that helps 😀

**Comment by vHanda** [comment]
> Supports 5 configurable options:
> 
>     1. File types to rename (comma separated list of extensions you want it to handle. png,jpg,gif are the defaults)
> 

What should happen when importing a 'jpg' if the jpg extension is not in the list?

>     2. Ignore folders (a new line separated list of folder names with relative paths to ignore even if they contain pattern matched extentions)
> 
>     3. Rename only linked attachments Boolean (if an image is not linked in the current note, it will not be renamed)

so this way one could add an "attachment" to the git repo, but the file's name wouldn't be touched?

> 
>     4. Save a previous name Boolean (when it changes the file's name, it uses the original name as the link's description
> 
>     5. Delete duplicates Boolean (if there are duplicate files in the same folder but with different names, it will delete the duplicates. If turned off, no file's will be renamed at all on this condition)

Hmm. The options will need a lot of explanation, someone may think that 'delete duplicates' works across the entire repo and not just the same folder.

Also, just to confirm, you would like these 5 + the ones I mentioned?

**Comment by majerich** [comment]
> > Supports 5 configurable options:
> > ```
> > 1. File types to rename (comma separated list of extensions you want it to handle. png,jpg,gif are the defaults)
> > ```
> 
> What should happen when importing a 'jpg' if the jpg extension is not in the list?
> 
> > ```
> > 2. Ignore folders (a new line separated list of folder names with relative paths to ignore even if they contain pattern matched extentions)
> > 
> > 3. Rename only linked attachments Boolean (if an image is not linked in the current note, it will not be renamed)
> > ```
> 
> so this way one could add an "attachment" to the git repo, but the file's name wouldn't be touched?
> 
> > ```
> > 4. Save a previous name Boolean (when it changes the file's name, it uses the original name as the link's description
> > 
> > 5. Delete duplicates Boolean (if there are duplicate files in the same folder but with different names, it will delete the duplicates. If turned off, no file's will be renamed at all on this condition)
> > ```
> 
> Hmm. The options will need a lot of explanation, someone may think that 'delete duplicates' works across the entire repo and not just the same folder.
> 
> Also, just to confirm, you would like these 5 + the ones I mentioned?

If jpg isn't in the list of extentions in the config window, nothing would happen to the file when it's imported. Between the first two configuration options, include extentions and exclude folders you can narrow it down so only very specific files in the appropriate location get renamed.

For config #3, the situation yopu present is correct, "attachment" files put into the repo not not specifically linked in a document will never get renamed. They would have to be [[ ]] linked in a note AND have the proper extentiomn AND be in non-ignored folder.

The 5th configuration option, tbh I've nbever touiched from its default state of off. It worries me a bit. I have no idea how it determines duplicates. I don't think it's actually that big of a necessity. I think in the average use case, we're attaching a few image to a daily journal entry. I don't think it's common for someone to be adding 10's of images into a note and creating a photo gallery where duplicate files would become an issue.

To be perfectly honest, I would be over the moon with just the first 3 configuration options. Plus, I would like to say I LOVE(!!!) making the file name its MD5 hash! That's actually what the Oblivion plug-in *should* be doing imho. The naming scenarios you mentioned plus the MD5 hash name option I think are great. Then the 3 configuration options could apply to all the naming patterns (include extention, exclude directory, do not touch unlinked files).

The more I think about it, the less use I see for the 3rd configuration option tbh. Does it really matter if the file is linked or unlinked? I'm guessing its purpouse is as a safety mechanism so you don't put a file into an attachments location and then forget to actually refer to it anywhere so it becomes an orphan. Personally for day to day use as a journal, that situation would never really occur I don't think. I think the norm is adding the file and it is automatically linked into the note. I haven't tried just putting a file into a GitJournal entry without actually having it linked. 

Maybe for starters just the foirst two option then (include extentions list and exclude directories (folders) list). Any other options would most likely be an edge case and be a separate enhancement to the core functionality.... or somthing lol.

**Comment by vHanda** [comment]
> To be perfectly honest, I would be over the moon with just the first 3 configuration options. Plus, I would like to say I LOVE(!!!) making the file name its MD5 hash! That's actually what the Oblivion plug-in _should_ be doing imho. The naming scenarios you mentioned plus the MD5 hash name option I think are great

Heh. The [Obsidian's Plugin Documentation](https://github.com/derwish-pro/obsidian-unique-attachments) is using the md5. It's why I decided to use md5 instead of any other scheme :)
